Add more unit tests for transformers

kamasylvia opened the pull request:
# Issue
Fix #17 
# Goals
- [x] Add `file` field to the output.
- [x] Get specific values from a nested object by nested keys.
- [x] Compare the transformated value with the resource value.
- [x] Sort transformed data by subtests and files.
- [x] Check files existance.

Commit:
Add "file" field to transformed data.

Commit:
Add "file" field to transformed data.

Commit:
Test the output value with the recourse file.

Add `get_nested_items` to get the recourse value.

Commit:
Test the output value with the recourse file.

Add `get_nested_items` to get the recourse value.

Commit:
Add issue-17-unit-test branch to .travis.yaml

Commit:
Add issue-17-unit-test branch to .travis.yaml

Commit:
Change function name.

Commit:
Change function name.

Commit:
Reformat

Commit:
Reformat

Commit:
Update for python 3.8

Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working

Commit:
Update for python 3.8

Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working

gmierz commented on .travis.yml:
```diff
@@ -8,6 +8,7 @@ branches:
     - starter
     - issue-9-travis-pytest
```
Remove these `issue*` branches from the travis config file.

gmierz commented on transformer.py:
```diff
@@ -90,8 +90,13 @@ def process(self, name):
                     trfmdata.extend(data)
                 else:
+                    for entry in data["data"]:
+                        entry.update({"file": file})
```
Rather than having these two paths, let's change this to this:
```
if type(data) != list:
   data = [data]

for entry in data:
   for ele in entry["data"]:
        ele.update({"file": file})

trfmdata.extend(data)
```

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
     expected_file = "testing/outputs/single_json_output.json"
     assert filecmp.cmp(actual_file, expected_file)
+
+    with open(actual_file, "r") as af:
```
This should be a second test. Also, you're only testing one transform, but you should be testing all of the transforms in this function. At the least, implement a test for the default transformer.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
-import pytest
+from utilities import get_values_from_nested_object
```
Project-specific imports should go below the python module imports.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
+            for data in entry["data"]:
+                assert "file" in data
+                assert "value" in data
+                assert "xaxis" in data
```
Check to make sure the number of `xaxis` entries equal to one value  (i.e. the number of data entries with xaxis == 1) is equal to the number of values you find in the file. In other words, `ret` from `get_values_from_nested_object` should have `len(ret) == len(num_xaxis_entries_with_value_1)`

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
+
+    with open(actual_file, "r") as af:
+        actual_data = json.load(af)
+        assert isinstance(actual_data, list)
```
Check that the list has the same number of file groupings that were defined in the config.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
+
+        for entry in actual_data:
+            assert "data" in entry
+            assert isinstance(entry["data"], list)
```
Check that the number of entries in the list is equal to the number of files times the number of page cycles (3 files X 3 page cycles = 9 entries)

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,36 @@
+        for entry in actual_data:
+            assert "data" in entry
+            assert isinstance(entry["data"], list)
+            assert "name" in entry
```
Check to make sure the name is in the file groupings keys.

gmierz commented on utilities.py:
```diff
@@ -53,3 +53,29 @@ def _helper(data, parent_dir, ret):
     return _helper(data, parent_dir, {})
+
+
+def get_values_from_nested_object(nested_obj, nested_keys):
```
This name is a bit long, can you change it to: `get_nested_values`

gmierz reviewed (changes requested):
Nice! There's a few more things to check in the tests.

Commit:
Add more tests.

Commit:
Add more tests.

Commit:
Add more check.

Commit:
Add more check.

Commit:
Add more check.

Commit:
Add more check.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,72 @@
+        with open(test_config_filepath, "r") as f, open(target_config_filepath, "r") as g:
+
+            test_config = yaml.safe_load(f)
+            target_config = yaml.safe_load(g)
```
You can get the data from these two files and immediately close the config files. That way you won't have all your other code in the indent.
```
        with open(test_config_filepath, "r") as f, open(target_config_filepath, "r") as g:
            test_config = yaml.safe_load(f)
            target_config = yaml.safe_load(g)
       assert "file_groups" in ...
```

You can do the same thing below when you open the test output.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,72 @@
+
+            test_output_filepath = test_ptnb.parse_output()
+            target_output_filepath = target_ptnb.parse_output()
+            assert filecmp.cmp(test_output_filepath, target_output_filepath)
```
If the files are different, this is going to fail immediately, and it won't give us much information about why. Move this to the end of the test.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,72 @@
+
+
+class TestTransformer(object):
+    def default_transformer(self, test_config_filepath, target_config_filepath):
```
You will have to add a lot of comments about what this function is doing in the code, and up here in a docstring.

gmierz commented on testing/tests/test_custom_single_json_retriever.py:
```diff
@@ -1,8 +1,72 @@
-    actual_file = "testing/outputs/single_json_output_test.json"
-    expected_file = "testing/outputs/single_json_output.json"
-    assert filecmp.cmp(actual_file, expected_file)
+    def test_single_json_retriever(self):
```
Add a second test to this class for the SimplePerfherderTransform now. Your default_transformer function should be able to check that transformation as well.

Also, can you change the name from `default_transformer` to `check_transformation`?

gmierz reviewed (changes requested):
Nice testing function! It shouldn't be difficult to add a test for the SimplePerfherderTransform now.

Commit:
Add more tests.

- Add test for SimplePerfherderTransformer.
- Add comments for check_transformation.

Commit:
Add more tests.

- Add test for SimplePerfherderTransformer.
- Add comments for check_transformation.

Commit:
Fix Travis test.

Commit:
Fix Travis test.

Commit:
Add file exists check.

Commit:
Add file exists check.

Commit:
Sort files.

Commit:
Sort files.

Commit:
Sort transformed data by subtest.

Commit:
Sort transformed data by subtest.

Commit:
Regenerate transformed output.

Commit:
Regenerate transformed output.

gmierz commented on .travis.yml:
```diff
@@ -6,13 +6,13 @@ branches:
     - starter
-    - issue-9-travis-pytest
-    - issue-11-add-Black-formatter
+    - issue-17-unit-tests
```
This needs to be removed.

gmierz commented on perftestnotebook.py:
```diff
@@ -152,6 +161,7 @@ def process(self):
                 all_results[func] = getattr(self.analyzer, func)()
             return all_results
 
+        fmt_data.sort(key=lambda entry: entry["subtest"])
```
This can be [done here](https://github.com/mozilla/mach-perftest-notebook-dev/blob/starter/perftestnotebook.py#L117), you can get a list of the subtest keys from the dict `files`, then sort it and iterate over that list instead of `files.items()`:
```
subtests = list(files.keys())

if self.sort_files:
    subtests.sort()

for subtest in subtests:
    proc_files = files[subtest]
    # You should change the files variable to `proc_files` to prevent any issues in the future with overriding variables.
```

gmierz commented on transformer.py:
```diff
@@ -75,6 +75,7 @@ def process(self, name):
         :return dict: Merged data.
         """
         trfmdata = []
+        self.files.sort()
```
It's better to do this sorting here since we don't want to let users override it: https://github.com/mozilla/mach-perftest-notebook-dev/blob/starter/perftestnotebook.py#L102

```
if self.sort_files:
    if type(files) == list:
        files.sort()
    else:
        for _, file_list in files:
            file_list.sort()
```

gmierz reviewed (changes requested):
Awesome! Almost there now. Be sure to add a command line flag `--sort-files` to enable sorting, since this is only really necessary for testing.

Commit:
Sort file dict.

Commit:
Sort file dict.

Commit:
Sort files.

Commit:
Sort files.

gmierz commented on testing/tests/test_transformers.py:
```diff
@@ -0,0 +1,105 @@
+            - Both of test and local config files from the arguments have "file_groups" field.
+            - Each entry of the transformed output has "data", "name", and "subtest" fields.
+                - Each "data" field has "file", "value", and "xaxis" fields.
+                    - Each file from "file" field is included in the "file_groups" field of the config file.
```
These lines seem to be longer than 100 columns, can you check if it passes black formatting?

kamasylvia replied:
It passes black formatting. I guess black does not format comments. Now I modify them to less than 100 columns manually.

gmierz reviewed (changes requested):
Nice!! Nearly there now, just a few comments left to resolve.

gmierz commented:
For the conflicts, you will need to rebase your branch on the latest changes in the starter branch and fix the conflicts as they are found.

Commit:
Update .gigignore

Commit:
Update .gigignore

Commit:
Merge remote-tracking branch 'origin/issue-17-unit-tests' into issue-17-unit-tests

Commit:
Add sort_fiels.

Commit:
Remove pipenv Pipfile

Commit:
Update perftestnotebook.py

Commit:
Add Black to requirements.

Commit:
Add --sort-files as arg.

Commit:
Update --sort-files.

Commit:
Update .travis.yml

Commit:
Update requirements.

gmierz commented on perftestnotebook.py:
```diff
@@ -216,7 +230,9 @@ def main():
 
-    ptnb = PerftestNotebook(config["file_groups"], config, custom_transform=custom_transform)
+    ptnb = PerftestNotebook(
+        config["file_groups"], config, custom_transform=custom_transform, sort_files=args.sort_files
```
Make a list for these arguments:
```
config['file_groups'],
config,
...
```

gmierz commented on testing/tests/test_transformers.py:
```diff
@@ -0,0 +1,118 @@
+    def check_transformation(self, test_config_filepath, target_config_filepath, get_nested_keys):
+        """
+        This function checks:
+            - Both of test and local config files from the arguments have "file_groups" field.
```
nit: Should be `That both testing and local config files from the arguments have "file_groups" field.`

Also, can you add some inline comments in the test to make where all these tests are happening more clear.

gmierz reviewed (changes requested):
I was gonna ask you remove the data prints but you're way ahead of me on that, thanks! :)

With these changes, I don't see anything else left so we're almost there.

gmierz commented on .travis.yml:
```diff
@@ -6,14 +6,13 @@ branches:
     - starter
-    - issue-9-travis-pytest
-    - issue-11-add-Black-formatter
+    - /.*/
```
Please remove this before landing btw. I think the reason you have to do this is because your original patch had it and after removing it, it won't run on the PR anymore because it's still trying to run on your branch.

Commit:
Add inline comments to test file.

gmierz reviewed (approved):
Excellent, thanks! :)
